File: 06-Arrays/3-arrays-2D.cpp
#include <iostream>

const int M = 5; // rows
const int N = 4; // columns

void fill(double data[][N], const int sizex, const int sizey);
void print(double data[][N], const int sizex, const int sizey);
void print(double data[][M], const int sizex, const int sizey);
void transpose(double data[][N], double dataT[][M], const int sizex, const int sizey);

int main(int argc , char **argv)
{
    // declare variables and matrix
    double matrix [M][N] {{0.0}};

    // fill
    fill(matrix, M, N);

    // transpose
    double matrixT [N][M] {{0.0}};
    transpose(matrix, matrixT, M, N);

    // print
    print(matrix, M, N);
    print(matrixT, N, M);

    return 0;
}

void fill(double data[][N], const int sizex, const int sizey)
{
    for (int ix = 0; ix < sizex; ++ix) {
        for (int iy = 0; iy < sizey; ++iy) {
            data[ix][iy] = ix*sizey + iy;
        }
    }
}

void print(double data[][N], const int sizex, const int sizey)
{
    for (int ix = 0; ix < sizex; ++ix) {
        for (int iy = 0; iy < sizey; ++iy) {
            std::cout << data[ix][iy] << "  ";
        }
        std::cout << "\n";
    }
}

void print(double data[][M], const int sizex, const int sizey)
{
    for (int ix = 0; ix < sizex; ++ix) {
        for (int iy = 0; iy < sizey; ++iy) {
            std::cout << data[ix][iy] << "  ";
        }
        std::cout << "\n";
    }
}

void transpose(double data[][N], double dataT[][M], const int sizex, const int sizey)
{
   for (int ix = 0; ix < sizex; ++ix) {
       for (int iy = 0; iy < sizey; ++iy) {
           dataT[iy][ix] = data[ix][iy];
       }
    }
}

File: 06-Arrays/3-arrays-2D.py
"""
Description: Arrays in 2D
"""
import numpy as np

N = 5 #rows
M = 4 #columns

#Two forms of initialize 2D arrays (the first one has to be filled)
matrix1 = np.zeros((N,M))
matrix2 = np.array([ [1,23,6],[8,7,5] ])

#To fill or access to the elements
matrix1[4][3] = 3
matrix1[2,2] = 2
a = matrix2[0,1]

def fill(array, row, col):
    for ix in np.arange(row):
        for iy in np.arange(col):
            array[ix][iy] = ix*col + iy

def transpose(array, Tarray, row, col):
    for ix in np.arange(row):
        for iy in np.arange(col):
            Tarray[iy][ix] = array[ix][iy]
            
#Fill original matrix
fill(matrix1, N, M)

#Initialize transpose
Tmatrix = np.zeros((M,N))

#Get the transpose
transpose(matrix1, Tmatrix, N, M)

#print both
print("Original matrix:\n", matrix1, '\n' , "Transpose:\n" , Tmatrix)





"""
//////////////////////////////////////////////////////////////////////////
Created on Fri Jan  1 13:32:11 2021                                    //
                                                                      //
@author: Nicolás Gómez                                               //
//////////////////////////////////////////////////////////////////////
"""

